src/error_diffusion.jl

```diff
@@ -14,147 +16,164 @@ julia> cs = ColorSchemes.PuOr_7.colors; # using ColorSchemes.jl for color palett
-    filter::F
+struct ErrorDiffusion{C,I,V} <: AbstractDither
+    inds::I # indices of filter
+    vals::V # values of filter
     clamp_error::Bool
 end
-function ErrorDiffusion(filter; color_space=XYZ, clamp_error=true)
-    return ErrorDiffusion{typeof(filter),float32(color_space)}(filter, clamp_error)
+function ErrorDiffusion(colorspace, inds, vals, clamp_error)
+    length(inds) != length(vals) &&
+        throw(ArgumentError("Lengths of filter indices and values don't match."))
+    return ErrorDiffusion{float32(colorspace),typeof(inds),typeof(vals)}(
+        inds, vals, clamp_error
+    )
+end
+function ErrorDiffusion(
+    filter::AbstractMatrix, offset::Integer, colorspace=XYZ; clamp_error=true
+)
+    require_one_based_indexing(filter)
+    CI = CartesianIndices(filter) .- CartesianIndex(1, offset)
+    mask = .!iszero.(filter) # only keep non-zero values
+    return ErrorDiffusion(colorspace, CI[mask], filter[mask], clamp_error)
 end
 
 function binarydither!(alg::ErrorDiffusion, out::GenericGrayImage, img::GenericGrayImage)
-    # this function does not yet support OffsetArray
+    # This function does not yet support OffsetArray
     require_one_based_indexing(img)
 
     # Change from normalized intensities to Float as error will get added!
-    # eagerly promote to the same type to make loop run faster
+    # Eagerly promote to the same type to make loop run faster.
     FT = floattype(eltype(img))
-    img = FT.(img)
-    filter = eltype(FT).(alg.filter)
+    img = convert.(FT, img)
 
-    drs = axes(alg.filter, 1)
-    dcs = axes(alg.filter, 2)
+    T = eltype(out)
+    T0, T1 = zero(T), oneunit(T)
 
-    FT0, FT1 = FT(0), FT(1)
+    vals = convert.(eltype(FT), alg.vals)
 
     @inbounds for r in axes(img, 1)
         for c in axes(img, 2)
-            px = img[r, c]
+            I = CartesianIndex(r, c)
+            px = img[I]
             alg.clamp_error && (px = clamp01(px))
 
-            px >= 0.5 ? (col = FT1) : (col = FT0) # round to closest color
-            out[r, c] = col # apply pixel to dither
-            err = px - col  # diffuse "error" to neighborhood in filter
-
-            for dr in drs
-                for dc in dcs
-                    if (r + dr) in axes(img, 1) && (c + dc) in axes(img, 2)
-                        img[r + dr, c + dc] += err * filter[dr, dc]
-                    end
-                end
-            end
+            out[I] = ifelse(px >= 0.5, T1, T0) # round to closest color
+            err = px - out[I]  # diffuse "error" to neighborhood in filter
+            diffuse_error!(img, err, I, alg.inds, vals)
         end
     end
-
     return out
 end
 
+# Diffuse error `err` in `img` around neighborhood of coordinate `I` defined by `filter`.
+function diffuse_error!(img, err, I, inds, vals)
+    for i in 1:length(inds)
+        N = I + inds[i] # index of neighbor
+        if checkbounds(Bool, img, N)
+            img[N] += err * vals[i]
+        end
+    end
+    return nothing
+end
```

Comment:
We can mark this as `Base.@propagate_inbounds` so that

```julia
    @inbounds diffuse_error!(img, err, I, alg.inds, vals)
```

skips bounds check.

The loop can be separated into two parts: 1) inner loop that no need to do boundscheck, and 2) boundary that boundscheck is necessary. This can be easily done using TiledIterations.jl, an example of this can be found in ImageMorphology https://github.com/JuliaImages/ImageMorphology.jl/blob/3181b69eab153fec9eeb0eaa55e97000252fa895/src/extreme_filter.jl#L147-L166